src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -38,117 +39,98 @@
  * users the opportunity to be able to provide config with a unit of their choice in cassandra.yaml as per the available
  * options. (CASSANDRA-15234)
  */
-public class DataStorageSpec
+public abstract class DataStorageSpec
 {
     /**
      * Immutable map that matches supported time units according to a provided smallest supported time unit
      */
     private static final ImmutableMap<DataStorageUnit, ImmutableSet<DataStorageUnit>> MAP_UNITS_PER_MIN_UNIT =
-    ImmutableMap.of(KIBIBYTES, ImmutableSet.of(KIBIBYTES, MEBIBYTES, GIBIBYTES),
+    ImmutableMap.of(BYTES, ImmutableSet.of(BYTES, KIBIBYTES, MEBIBYTES, GIBIBYTES),
+                    KIBIBYTES, ImmutableSet.of(KIBIBYTES, MEBIBYTES, GIBIBYTES),
                     MEBIBYTES, ImmutableSet.of(MEBIBYTES, GIBIBYTES));
     /**
      * The Regexp used to parse the storage provided as String.
      */
-    private static final Pattern STORAGE_UNITS_PATTERN = Pattern.compile("^(\\d+)(GiB|MiB|KiB|B)$");
+    private static final Pattern UNITS_PATTERN = Pattern.compile("^(\\d+)(GiB|MiB|KiB|B)$");
 
     private final long quantity;
 
     private final DataStorageUnit unit;
 
-    public DataStorageSpec(String value)
+    private DataStorageSpec(long quantity, DataStorageUnit unit, DataStorageUnit smallestUnit, long max)
     {
-        //parse the string field value
-        Matcher matcher = STORAGE_UNITS_PATTERN.matcher(value);
-
-        if (!matcher.find())
-        {
-            throw new ConfigurationException("Invalid data storage: " + value + " Accepted units: B, KiB, MiB, GiB" +
-                                             " where case matters and only non-negative values are accepted");
-        }
-
-        quantity = Long.parseLong(matcher.group(1));
-        unit = DataStorageUnit.fromSymbol(matcher.group(2));
-    }
-
-    DataStorageSpec(long quantity, DataStorageUnit unit)
-    {
-        if (quantity < 0)
-            throw new ConfigurationException("Invalid data storage: value must be positive, but was " + quantity);
-
         this.quantity = quantity;
         this.unit = unit;
+
+        validateQuantity(quantity, unit, smallestUnit, max);
     }
 
-    public DataStorageSpec (String value, DataStorageUnit minUnit)
+    private DataStorageSpec (String value, DataStorageUnit minUnit)
```

Comment:
Nit: additional whitespace:
```suggestion
    private DataStorageSpec(String value, DataStorageUnit minUnit)
```